### PR TITLE
GG-19486 IgniteOutOfMemoryException should not prevent cache clearing

### PR DESCRIPTION
**Pull request by sk0x50**

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/IgniteCacheDatabaseSharedManager.java [hunk 1]
```diff
@@ -982,6 +985,63 @@ public void releaseHistoryForPreloading() {
...
+     * For example, removing a row from underlying store may require allocating a new data page
+     * in order to move a tracked page from one bucket to another one which does not have a free space for a new stripe.
+     * See {@link AbstractFreeList#removeDataRowByLink}.
+     * Therefore, inserting a new entry should be prevented in case of some threshold is exceeded.
+     *
+     * @param region Data region to be checked.
+     * @throws IgniteOutOfMemoryException In case of the given data region does not have enough free space
+     * for putting a new entry.
+     */
+    public void ensureFreeSpaceForInsert(DataRegion region)throws IgniteOutOfMemoryException {
```
Comment (glukos) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/IgniteCacheDatabaseSharedManager.java]: space between ) and throws

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/IgniteCacheDatabaseSharedManager.java [hunk 2]
```diff
@@ -982,6 +985,63 @@ public void releaseHistoryForPreloading() {
...
+
+        long memorySize = regCfg.getMaxSize();
+
+        PageMemory pageMem = region.pageMemory();
+
+        int sysPageSize = pageMem.systemPageSize();
+
+        CacheFreeList freeList = freeListMap.get(regCfg.getName());
+
+        long dirtyPages = (pageMem.loadedPages() - freeList.emptyDataPages());
```
Comment (glukos) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/IgniteCacheDatabaseSharedManager.java]: I guess, "dirty" isn't the best wording for this number as it clashes with "dirty page" abstraction from persistence

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/IgniteCacheDatabaseSharedManager.java [hunk 3]
```diff
@@ -982,6 +985,63 @@ public void releaseHistoryForPreloading() {
...
+
+        boolean oomThreshold = (memorySize / sysPageSize) <
+            (dirtyPages * (8.0 /*link size*/ / sysPageSize + 1) + 256 /*one page per bucket*/);
```
Comment (glukos) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/IgniteCacheDatabaseSharedManager.java]: A bit risky.
Less than sysPageSize space is used in PagesListMetaIO / PagesListNodeIO (minus headers and so on) for storing links. I'd multiply it by 1.5 to be completely sure.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/IgniteCacheDatabaseSharedManager.java (hunk 2 above)
Comment (sk0x50) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/IgniteCacheDatabaseSharedManager.java]: Agreed. I will change it to `nonEmptyPages`.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/IgniteCacheDatabaseSharedManager.java (hunk 3 above)
Comment (sk0x50) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/IgniteCacheDatabaseSharedManager.java]: You are right. The space that can be used for storing links should be calculated as follows, I think:
`pageMem.pageSize() - PagesListNodeIO.PAGE_IDS_OFF`.
Perhaps, a constant multiplier would be a good approach in order to eliminate different versions of `PagesListNodeIO/PagesListMetaIO` (that can be added in the future).

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/IgniteCacheDatabaseSharedManager.java (hunk 1 above)
Comment (sk0x50) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/IgniteCacheDatabaseSharedManager.java]: Fixed

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/IgniteCacheDatabaseSharedManager.java (hunk 3 above)
Comment (sk0x50) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/IgniteCacheDatabaseSharedManager.java]: I updated the calculation as follows:
```java
        // The maximum number of pages that can be allocated (memorySize / systemPageSize)
        // should be greater or equal to the current number of non-empty pages plus
        // the number of pages that may be required in order to move all pages to a reuse bucket,
        // that is equal to nonEmptyPages * 8 / pageSize, where 8 is the size of a link.
        // Note that not the whole page can be used to storing links,
        // see PagesListNodeIO and PagesListMetaIO#getCapacity(), so we pessimistically multiply the result on 1.5,
        // in any way, the number of required pages is less than 1 percent.
        boolean oomThreshold = (memorySize / pageMem.systemPageSize()) <
            (nonEmptyPages * (8.0 / pageMem.pageSize() + 1) * 1.5 + 256 /*one page per bucket*/);
```

**Pull request closed without merging**